sw/device/silicon_creator/rom/e2e/rom_e2e_ret_ram_keep_test.c

```diff
@@ -0,0 +1,114 @@
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+#include "sw/device/silicon_creator/lib/drivers/retention_sram.h"
+#include "sw/device/silicon_creator/lib/drivers/rstmgr.h"
+#include "sw/device/silicon_creator/lib/test_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+OTTF_DEFINE_TEST_CONFIG();
+
+enum {
+  kPattern = 0xab,
+};
+
+/**
+ * Check that the values in the retention SRAM have not changed
+ * from kPattern. The 0th word of the SRAM is exempt as it is
+ * used to store the reset reason.
+ */
+bool check_ram_unchanged(retention_sram_t *ret) {
+  LOG_INFO("Checking that retention SRAM values are unchanged");
+  static uint32_t raw[sizeof(retention_sram_t) / sizeof(uint32_t)];
+  uint32_t pattern32;
+  memset(&pattern32, kPattern, sizeof(pattern32));
+
+  // Ensure that all written sections were saved
+  memcpy(raw, ret, sizeof(retention_sram_t));
+
+  bool unchanged = true;
+  // Skip checking index 0 since that is used to store the
+  // reset reason and will be changed.
+  for (size_t i = 1; i < ARRAYSIZE(raw); ++i) {
+    if (raw[i] != pattern32) {
+      LOG_ERROR("Retention SRAM changed at offset %u (%x --> %x).", i,
+                pattern32, raw[i]);
+      unchanged = false;
+    }
+  }
+  return unchanged;
+}
+
+rom_error_t retention_ram_keep_test(void) {
+  // Variables of type `retention_sram_t` are static to reduce stack usage.
+  static retention_sram_t *ret;
+  ret = retention_sram_get();
+  uint32_t reset_reasons = ret->reset_reasons;
+
+  // Verify that reset_reasons reports POR.
+  if (bitfield_bit32_read(reset_reasons, kRstmgrReasonPowerOn)) {
+    // This branch runs after the POR after initializing the testing environment
+
+    LOG_INFO("Writing known pattern to ret RAM");
+    memset(ret, kPattern, sizeof(retention_sram_t));
+
+    // Initialize pwrmgr
+    dif_pwrmgr_t pwrmgr;
+    CHECK_DIF_OK(dif_pwrmgr_init(
+        mmio_region_from_addr(TOP_EARLGREY_PWRMGR_AON_BASE_ADDR), &pwrmgr));
+
+    // Initialize aon timer
+    // Issue a wakeup signal in ~150us through the AON timer.
+    //
+    // At 200kHz, threshold of 30 is equal to 150us. There is an additional
+    // ~4 cycle overhead for the CSR value to synchronize with the AON clock.
+    // We should expect the wake up to trigger in ~170us. This is sufficient
+    // time to allow pwrmgr config and the low power entry on WFI to complete.
+    //
+    // Adjust the threshold for Verilator since it runs on different clock
+    // frequencies.
+    uint32_t wakeup_threshold = kDeviceType == kDeviceSimVerilator ? 300 : 30;
+
+    dif_aon_timer_t aon_timer;
+    CHECK_DIF_OK(dif_aon_timer_init(
+        mmio_region_from_addr(TOP_EARLGREY_AON_TIMER_AON_BASE_ADDR),
+        &aon_timer));
+    aon_timer_testutils_wakeup_config(&aon_timer, wakeup_threshold);
+
+    // Enter low-power
+    pwrmgr_testutils_enable_low_power(&pwrmgr,
+                                      kDifPwrmgrWakeupRequestSourceFive, 0);
```

Comment:
I got this `kDifPwrmgrWakeupRequestSourceFive` value from pwrmgr_smoketest.c, but I'm not entirely clear on what it's doing. Is this the correct way to perform a low-power entry?

Reply:
Correct, thanks for checking. Would you mind adding the following assertion (from DIF library [here](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/lib/dif/dif_pwrmgr.c;drc=e500fca1e37f6e2a690203f117b4392352ed6367;l=84)) for sake of clarity?

Reply:
Done